Update templates with bearer authentication support for BYO server cases

thepetk opened the pull request:
### What does this PR do?:
This is the final PR for DEVAI-157 & DEVAI-158. It:

* Pulls the latest gitops updates introduced by https://github.com/redhat-ai-dev/ai-lab-app/pull/28
* Pulls the latest sample updates introduced by https://github.com/redhat-ai-dev/ai-lab-samples/pull/13
* Updates all templates with the all fields related to bearer authentication for BYO server cases.

I've marked the PR as draft because I think is fair to do a final round of tests now that all related PRs are merged.

### Which issue(s) this PR fixes:
https://issues.redhat.com/browse/DEVAI-158

### PR acceptance criteria:
Testing and documentation do not need to be complete in order for this PR to be approved. We just need to ensure tracking issues are opened and linked to this PR, if they are not in the PR scope due to various constraints.

- [ ] Tested and Verified

  <!-- _I have verified that the changes were tested manually_ -->

- [ ] Documentation (READMEs, Product Docs, Blogs, Education Modules, etc.)

   <!-- _This includes READMEs, Product Docs, Blogs, Education Modules, etc._ -->


### How to test changes / Special notes to the reviewer:
For `chatbot`, `codegen` & `rag` you can use one of the already deployed models and try to authenticate with your app-key.

For `audio-to-text` & `object-detection` please check this guide here: https://github.com/redhat-ai-dev/ai-lab-app/pull/28#issuecomment-2399489053

maysunfaisal commented on skeleton/template.yaml:
```diff
@@ -80,12 +80,36 @@ spec:
+                            ui:help: Please provide the name of the secret that your bearer token is stored.
+                            type: string
+                          modelEndpointSecretKey:
+                            title: Model Enpoint Secret Key
```
```suggestion
                            title: Model Endpoint Secret Key
```

maysunfaisal commented on skeleton/template.yaml:
```diff
@@ -80,12 +80,36 @@ spec:
+                        properties:
+                          modelEndpointSecretName:
+                            title: Model Enpoint Secret Name
+                            ui:help: Please provide the name of the secret that your bearer token is stored.
```
```suggestion
                            ui:help: Provide the name of the Secret containing your bearer token.
```

maysunfaisal commented on skeleton/template.yaml:
```diff
@@ -80,12 +80,36 @@ spec:
+                      then:
+                        properties:
+                          modelEndpointSecretName:
+                            title: Model Enpoint Secret Name
```
```suggestion
                            title: Model Endpoint Secret Name
```

maysunfaisal commented on skeleton/template.yaml:
```diff
@@ -80,12 +80,36 @@ spec:
+                            type: string
+                          modelEndpointSecretKey:
+                            title: Model Enpoint Secret Key
+                            ui:help: Please provide the key of the secret that the token value is assigned to.
```
```suggestion
                            ui:help: Provide the Secret's key containing the token value.
```

Jdubrick replied:
IMO this should probably be `Provide the key of the Secret containing the token value.`

maysunfaisal commented:
@thepetk I dont have a running cluster, can you post screenshot of the template fields?

thepetk commented:
> @thepetk I dont have a running cluster, can you post screenshot of the template fields?

@maysunfaisal @Jdubrick FYI I'm currently working on providing screenshots for the UX to review to get their input :heart:

Jdubrick commented on skeleton/template.yaml:
```diff
@@ -80,12 +80,36 @@ spec:
+                            type: string
+                          modelEndpointSecretKey:
+                            title: Model Endpoint Secret Key
+                            ui:help: Provide the Secret's key containing the token value.
```
```suggestion
                            ui:help: Provide the key of the Secret containing the token value.
```

maysunfaisal commented on skeleton/template.yaml:
```diff
@@ -80,12 +80,36 @@ spec:
+                        properties:
+                          modelEndpointSecretName:
+                            title: Model Endpoint Secret Name
+                            ui:help: Provide the name of the Secret containing your bearer token.
```
Have you tried this secret reference in a custom namespace? 🤔 

In the templates wizard,

1. We ask for the Namespace to be deployed
2. Deployment references the Secret in the same Namespace as Deployment https://github.com/redhat-ai-dev/ai-lab-template/pull/40/files#diff-f3f000f5975fd959a1fd26857e77e8fcc8c892cc11f8cd83bca93e5b030b7591R39 
3. Default Namespace is fine, but if user is providing a custom Namespace via Step 1; they will have to create the Namespace before creating the Secret?

thepetk replied:
Yes, you need to have the secret created in your namespace upfront in order to provide it here.

> Default Namespace is fine, but if user is providing a custom Namespace via Step 1; they will have to create the Namespace before creating the Secret?

I think we don't create the default namespace (meaning the `ai-rhdh-app-development`) when we run the installer right? So for every case the namespace and the secret should be there beforehand. Otherwise I'm not sure how we can avoid passing the token itself in the wizzard. Might be better if we capture that in the `ui:help`?

Jdubrick replied:
Currently `ai-rhdh-app-development`, `ai-rhdh-app-stage` and `ai-rhdh-app-prod` are created when the installer is run but we also added the Tekton tasks that allow you to enter any namespace in the template and it sets it up for you

thepetk replied:
> Currently ai-rhdh-app-development, ai-rhdh-app-stage and ai-rhdh-app-prod are created when the installer is run but we also added the Tekton tasks that allow you to enter any namespace in the template and it sets it up for you

Yeah tbh as the installer is in progress I wasn't exactly sure how it's done. But in general I think the principal is the same. We need to define the secret prior to the template creation.

I agree with @maysunfaisal and @johnmcollier I'll try adding more context in the `ui:help field`.

thepetk replied:
Tried to add more info here: [af11a9e](https://github.com/redhat-ai-dev/ai-lab-template/pull/40/commits/af11a9ea05ac40c13233f6d23faebc76b1784687)

maysunfaisal commented:
@thepetk Can the model name field be moved after the Secret name and Key name references?

thepetk commented:
> @thepetk Can the model name field be moved after the Secret name and Key name references?

yeah I think is a good point. I think it can and I'll try tomorow :)

thepetk commented:
> > @thepetk Can the model name field be moved after the Secret name and Key name references?
> 
> yeah I think is a good point. I think it can and I'll try tomorow :)

It should be ok in `[53e5288d](https://github.com/redhat-ai-dev/ai-lab-template/pull/40/commits/789d04be0ed4afc1096fc3af55b4ed1a53e5288d)`. I'm in the process of testing everything atm, I'll share screenshots in a bit.

One very weird case I've seen is I was unable to create a template coming from a branch that in the name has `/`. More detailed if you try to import the template from my fork's branch the first step of RHDH should fail (`fetch base`) because github API gives 404.

Jdubrick commented:
> > > @thepetk Can the model name field be moved after the Secret name and Key name references?
> > 
> > 
> > yeah I think is a good point. I think it can and I'll try tomorow :)
> 
> It should be ok in `[53e5288d](https://github.com/redhat-ai-dev/ai-lab-template/pull/40/commits/789d04be0ed4afc1096fc3af55b4ed1a53e5288d)`. I'm in the process of testing everything atm, I'll share screenshots in a bit.
> 
> One very weird case I've seen is I was unable to create a template coming from a branch that in the name has `/` (e.g my branch name is `feature/add_byo_server_token_auth_support`). More detailed if you try to import the template from my fork's branch the first step of RHDH should fail (`fetch base`) because github API gives 404.

For what it is worth I have been experiencing that issue with branch names containing `/` since we started working with RHDH, I think it is a limitation of the parser for URLs

thepetk commented:
> > > > @thepetk Can the model name field be moved after the Secret name and Key name references?
> > > 
> > > 
> > > yeah I think is a good point. I think it can and I'll try tomorow :)
> > 
> > 
> > It should be ok in `[53e5288d](https://github.com/redhat-ai-dev/ai-lab-template/pull/40/commits/789d04be0ed4afc1096fc3af55b4ed1a53e5288d)`. I'm in the process of testing everything atm, I'll share screenshots in a bit.
> > One very weird case I've seen is I was unable to create a template coming from a branch that in the name has `/` (e.g my branch name is `feature/add_byo_server_token_auth_support`). More detailed if you try to import the template from my fork's branch the first step of RHDH should fail (`fetch base`) because github API gives 404.
> 
> For what it is worth I have been experiencing that issue with branch names containing `/` since we started working with RHDH, I think it is a limitation of the parser for URLs

Do you know if it's a known issue in RHDH? Maybe we should report it.

Jdubrick commented:
@thepetk https://github.com/backstage/backstage/issues/2815 found this

thepetk commented:
> @thepetk [backstage/backstage#2815](https://github.com/backstage/backstage/issues/2815) found this

I think that's sufficient too :)

johnmcollier commented on templates/audio-to-text/template.yaml:
```diff
@@ -57,6 +57,30 @@ spec:
+                        properties:
+                          modelEndpointSecretName:
+                            title: Model Endpoint Secret Name
+                            ui:help: Provide the name of the Secret containing your bearer token.
```
In a similar vein to @maysunfaisal's comment, since the secret must be in the same namespace as the deployment namespace, it might be good to highlight that requirement here.

thepetk replied:
+1 (provided more context [here](https://github.com/redhat-ai-dev/ai-lab-template/pull/40#discussion_r1803619385))

johnmcollier commented:
@thepetk Should this be moved out of draft?

MariaLeonova commented:
I have the flow laid out in Miro step by step here: [Miro - Software Templates](https://miro.com/app/board/uXjVKt3DIE0=/?moveToWidget=3458764601871701435&cot=14 ) and am working on some UX recommendations.

MariaLeonova commented:
Couple UX recommendations:
 - There's quite a bit of pre-work required in Openshift, which is not directly referenced in the Software Templates wizard UI, and I would guess that would have to be covered in documentation.
 - I wouldn't address that ^ issue as part of this PR, but I would recommend to give it some consideration - if anything can be done to guide users better through the UI, between RHDH and Openshift.
 - my main concern is the clarity of the language, so I made some suggestions as to how try and make it clearer for the user what they are supposed to do:
 
![BYO Server-bearer2](https://github.com/user-attachments/assets/383ae29f-27ed-4ac4-b218-eeedf277349c)
  
- I also moved Model Name field to the bottom, assuming the secret name and key will be directly relevant to the Model Server.
- I am proposing slight changes to formatting: indentation of the Bearer credentials fields to the right, so it's clear they are a subsection of the model server endpoint fields. 

I put this screen in [Miro here](https://miro.com/app/board/uXjVKt3DIE0=/?moveToWidget=3458764603601099493&cot=14), so please drop comments and stickies on the mock directly to address particular bits of the UI.

thepetk commented:
> @thepetk Should this be moved out of draft?

@johnmcollier Yeap! I think now is ready for review. I've rebased my PR to the updates coming from https://github.com/redhat-ai-dev/ai-lab-template/pull/42 and now we can test the current updates with `ai-rhdh-installer`

thepetk commented:
> * There's quite a bit of pre-work required in Openshift, which is not directly referenced in the Software Templates wizard UI, and I would guess that would have to be covered in documentation.
> * I wouldn't address that ^ issue as part of this PR, but I would recommend to give it some consideration - if anything can be done to guide users better through the UI, between RHDH and Openshift.

@MariaLeonova I've tried to add some documentation here: [5165cd8](https://github.com/redhat-ai-dev/ai-lab-template/pull/40/commits/5165cd8d31061fbe1e71242f5cd716f3bc954d71)

> My main concern is the clarity of the language, so I made some suggestions as to how try and make it clearer for the user what they are supposed to do.

I've added your suggestions to the template. Below you can find the new screens:

![Screenshot from 2024-10-17 11-21-15](https://github.com/user-attachments/assets/3a991943-298f-4621-bebc-d204cfdd1182)

![Screenshot from 2024-10-17 11-21-25](https://github.com/user-attachments/assets/bbf16ef9-a5c8-4c68-9359-1ee7b75186ea)

> I also moved Model Name field to the bottom, assuming the secret name and key will be directly relevant to the Model Server.
> I am proposing slight changes to formatting: indentation of the Bearer credentials fields to the right, so it's clear they are a subsection of the model server endpoint fields.

Here I don't think is possible atm to have indentation in UI fields (cc @maysunfaisal correct me if I'm wrong). So I think better to keep the model name for the time being above the bearer token secion? WDYT?

Note, based on your feedback I've updated the names of the fields:
* Model Endpoint Secret Name -> Model Server Endpoint Secret Name
* Model Endpoint Secret Key -> Model Server Endpoint Secret Key
Not sure if I should update it to secret name and secret key now that indentation is not supported.

MariaLeonova commented:
I would suggest making the question next to the question mark a full sentence: `Is bearer authentication required?`

thepetk commented:
> ` [82a0d90](/redhat-ai-dev/ai-lab-template/pull/40/commits/82a0d90c77d4c8fe2b7111a6584adb808ebe46da)`

@MariaLeonova addressed here: [82a0d90](https://github.com/redhat-ai-dev/ai-lab-template/pull/40/commits/82a0d90c77d4c8fe2b7111a6584adb808ebe46da)

johnmcollier reviewed (dismissed):
Looks good from my end

maysunfaisal commented on skeleton/template.yaml:
```diff
@@ -86,6 +86,31 @@ spec:
+                  title: Is bearer authentication required?
+                  type: boolean
+                  default: false
+                  ui:help: Create a secret containing the authentication bearer in the preffered targeted Namespace first.
```
```suggestion
                  ui:help: Create a secret containing the authentication bearer in the preferred targeted Namespace first.
```

maysunfaisal replied:
Do we want to capitalize Secret? It is a Kubernetes resource after all like Namespace..

thepetk replied:
For me makes sense. @MariaLeonova wdyt?

thepetk replied:
I think the only unresolved item of this PR is if we'd like to use `Secret` instead of `secret`. Once this is resolved I'll merge it :)

cc @MariaLeonova @maysunfaisal

MariaLeonova replied:
@thepetk I agree with @maysunfaisal, let's capitalize to be consistent with other Kubernetes resources.
https://kubernetes.io/docs/concepts/configuration/secret/.

thepetk replied:
Addressed here: [86b2275](https://github.com/redhat-ai-dev/ai-lab-template/pull/40/commits/86b227512bbab1040e3b2b01e138c341f0939e1d)

maysunfaisal commented on skeleton/gitops-template/components/http/base/rhoai/initialize-dsp.yaml:
```diff
@@ -90,6 +90,13 @@ spec:
+                      - name: MODEL_ENDPOINT_BEARER
+                        valueFrom:
+                          secretKeyRef:
+                            name: $${{   values.modelEndpointSecretName   }}
```
Application pod:
```
env:
        - name: MODEL_ENDPOINT_BEARER
          valueFrom:
            secretKeyRef:
              name: ibm-granite-8b-app
              key: api_key
```

Notebook Pod:
```
env:
            - name: MODEL_ENDPOINT_BEARER
              valueFrom:
                secretKeyRef:
                  name: '-granite-8b-app'
                  key: ''
```
and this is the error on RHOAI UI:
<img width="534" alt="Screenshot 2024-10-17 at 3 16 35 PM" src="https://github.com/user-attachments/assets/4258eb45-738c-4e06-b535-b01cb1d93b59">

I confirmed that I entered the secret and key correctly:
<img width="575" alt="Screenshot 2024-10-17 at 3 14 12 PM" src="https://github.com/user-attachments/assets/c9dce446-03c4-4141-bda5-bcdfd383a64d">

The secret name is getting truncated? Also, the key is empty! 🤔

maysunfaisal replied:
Is it because of the double `$$` and the extra whitespaces within `{{ }}`?

thepetk replied:
Hmm, good catch.. I'm trying to reproduce it too because I didn't see that in the past (might be a customization made from my side).

I've added a fix here: https://github.com/redhat-ai-dev/ai-lab-app/pull/33 so the `modelEndpointSecretName` and `modelEndpointSecretKey` follow the same format with all the other values we use in the `initialize-dsp`.

thepetk replied:
I've added a `rag4` in my fork's main branch with this change and for me it works with this update.

* You can add the secret in a namespace of your preference.
* Import https://github.com/thepetk/ai-lab-template/blob/main/templates/rag4/template.yaml

![Screenshot from 2024-10-18 13-04-40](https://github.com/user-attachments/assets/67e2f026-8976-4b36-a292-07a3cb84001d)

thepetk replied:
addressed in: [7616e23](https://github.com/redhat-ai-dev/ai-lab-template/pull/40/commits/7616e23c39f55d7dcdfaa0678c8a77c0784bb1b7)

maysunfaisal replied:
<img width="1217" alt="Screenshot 2024-10-18 at 4 19 28 PM" src="https://github.com/user-attachments/assets/2d7f4740-91de-4a6c-b833-2247659b9ec8">

maysunfaisal reviewed (dismissed):
/lgtm 

I tried the RAG Template with RHOAI and they seem fine. I havent tested the audio/object template yet as I have to set up the env but I can do that later.

thepetk commented:
> /lgtm
> 
> I tried the RAG Template with RHOAI and they seem fine. I havent tested the audio/object template yet as I have to set up the env but I can do that later.

Thanks for the extra testing. From my side I was able to ran both `object-detection` & `audio-to-text`. Note for `audio-to-text` I've used an slight edited template [here](https://github.com/thepetk/ai-lab-template/blob/main/templates/audio-to-text-2/template.yaml) as the original sample cannot accept insecure connections ([see here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/audio/audio_to_text/app/whisper_client.py)).

maysunfaisal reviewed (approved):
/lgtm

is it just rebasing? @thepetk

thepetk commented:
> /lgtm
> 
> is it just rebasing? @thepetk

I've fixed the "secret" to "Secret", pulled latest `ai-lab-app` changes (a minor renaming) and generated all new templates

maysunfaisal commented:
@thepetk what was the difference in audio template for insecure connections? I couldnt see the difference in the link you provided..

thepetk commented:
> @thepetk what was the difference in audio template for insecure connections? I couldnt see the difference in the link you provided..

In case we'd like to allow insecure connections, we should add an extra kwarg `verify=False` [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/audio/audio_to_text/app/whisper_client.py#L21). This will avoid raising an exception for such cases. An example is [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/computer_vision/object_detection/app/object_detection_client.py#L35).

maysunfaisal commented:
> > @thepetk what was the difference in audio template for insecure connections? I couldnt see the difference in the link you provided..
> 
> In case we'd like to allow insecure connections, we should add an extra kwarg `verify=False` [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/audio/audio_to_text/app/whisper_client.py#L21). This will avoid raising an exception for such cases. An example is [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/computer_vision/object_detection/app/object_detection_client.py#L35).

@thepetk should it be in the ai lab recipe main repo then? 🤔

thepetk commented:
> > > @thepetk what was the difference in audio template for insecure connections? I couldnt see the difference in the link you provided..
> > 
> > 
> > In case we'd like to allow insecure connections, we should add an extra kwarg `verify=False` [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/audio/audio_to_text/app/whisper_client.py#L21). This will avoid raising an exception for such cases. An example is [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/computer_vision/object_detection/app/object_detection_client.py#L35).
> 
> @thepetk should it be in the ai lab recipe main repo then? 🤔

Not sure tbh, IMO it should be addressed as a separate issue. Hardcoding this option will potentially have an impact on everyone using this sample. Might be better to handle this value as env var having secure connections-only, enabled by default.

thepetk commented:
> > > @thepetk what was the difference in audio template for insecure connections? I couldnt see the difference in the link you provided..
> > 
> > 
> > In case we'd like to allow insecure connections, we should add an extra kwarg `verify=False` [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/audio/audio_to_text/app/whisper_client.py#L21). This will avoid raising an exception for such cases. An example is [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/computer_vision/object_detection/app/object_detection_client.py#L35).
> 
> but object template is also hardcoded, so that would also need to be updated hmm

Exactly and IMO finding a more secure setup for `object-detection` & handling secure/insecure connections in a consistent way is more important. That's also the main reason I believe it's out of the scope of DEVAI-158. That said better to create a new issue for this.

`audio-to-text` & `object-detection` were set like this before the JWT support and the samples were only altered for testing purposes (e.g. for `audio-to-text` I was only able to create a cluster with self-signed cert).

thepetk commented:
> > > > @thepetk what was the difference in audio template for insecure connections? I couldnt see the difference in the link you provided..
> > > 
> > > 
> > > In case we'd like to allow insecure connections, we should add an extra kwarg `verify=False` [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/audio/audio_to_text/app/whisper_client.py#L21). This will avoid raising an exception for such cases. An example is [here](https://github.com/containers/ai-lab-recipes/blob/main/recipes/computer_vision/object_detection/app/object_detection_client.py#L35).
> > 
> > 
> > but object template is also hardcoded, so that would also need to be updated hmm
> 
> Exactly and IMO finding a more secure setup for `object-detection` & handling secure/insecure connections in a consistent way is more important. That's also the main reason I believe it's out of the scope of DEVAI-158. Meaning, `audio-to-text` & `object-detection` were set like this before the JWT support and the samples were only altered for testing purposes (e.g. for `audio-to-text` I was only able to create a cluster with self-signed cert).That said better if I create a new issue for this.

Opened https://issues.redhat.com/browse/DEVAI-201
